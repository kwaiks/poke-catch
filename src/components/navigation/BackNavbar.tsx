import { css } from "@emotion/css";
import BackIcon from "assets/icons/Back";
import { useRouter } from "next/router"
import { useEffect, useState } from "react";
import { centerContentClass } from "styles/emotion/content";
import { navBarStyle } from "styles/emotion/list";

const backButtonWrapperClass = css`
    height: 30px;
    width: 30px;
    border-radius: 99px;
    ${centerContentClass};
`

export default function BackNavbar({safeHeight=200}:{safeHeight?:number}){
    const router = useRouter();
    const [overlap, setOverlap] = useState(false);

    useEffect(()=>{
        const scroll = () => {
            setOverlap(window.pageYOffset >= safeHeight)
        }
        if(window) {
            window.addEventListener("scroll", scroll);
            return () => {
                window.removeEventListener("scroll", scroll);
            }
        }
    },[])

    return (
        <nav data-testid="back-navbar" className={css`${navBarStyle}; background-color: ${overlap? "white" : "transparent"}`}>
            <div 
            data-testid="back-navbar-icon-wrapper"
            onClick={()=>router.back()}
            className={css`${backButtonWrapperClass}; background-color: ${overlap? "transparent" : "#f5f5f5"};`}>
                <BackIcon 
                />
            </div>
        </nav>
    )
}