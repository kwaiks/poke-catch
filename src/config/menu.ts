import HomeIcon from "../assets/icons/Home";
import PokeBallIcon from "../assets/icons/Pokeball";

const menu = [
    {
        Icon: HomeIcon,
        title: "Home",
        path: "/"
    },
    {
        Icon: PokeBallIcon,
        title: "My Pokemon",
        path: "/my-pokemon"
    },
];


export default menu;