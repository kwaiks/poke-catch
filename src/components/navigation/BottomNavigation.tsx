import { css } from "@emotion/css";
import { MAIN_COLOR } from "config/constant";
import menu from "config/menu";
import Link from "next/link";
import { useRouter } from "next/router";
import { fixedBottomClass } from "styles/emotion/content";

const footerStyle = css`
    ${fixedBottomClass}
    display: flex;
`;

const footerItemStyle = css`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    font-size: 10px;
    padding: 6px 0;
    width: 100%;
`;

export default function BottomNavigation(){
    const router = useRouter()

    return (
        <div 
        data-testid="bottom-nav"
        className={footerStyle} id="footer">
            {
                menu.map(({Icon, title, path}:any, idx)=>
                    <Link key={path} href={path} passHref>
                        <div 
                        data-testid={`nav-item-${title}`}
                        key={idx}
                        className={footerItemStyle}>
                            <Icon color={router.pathname === path ? MAIN_COLOR : "#000"}/>
                            <span>{title}</span>
                        </div>
                    </Link>
                )}
        </div>
    )
}