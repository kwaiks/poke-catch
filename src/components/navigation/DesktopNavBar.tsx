import { css } from "@emotion/css";
import { navBarStyle } from "styles/emotion/list";
import Image from "next/image";
import { MAIN_COLOR } from "config/constant";
import menu from "config/menu";
import { centerContentClass } from "styles/emotion/content";
import Link from "next/link";

const mainNavBarStyle = css`
    color: white;
    background-color: ${MAIN_COLOR};
    justify-content: space-between;
    padding: 0 10vw;
`;


export default function DesktopNavBar() {
    return (
        <nav 
        data-testid="desktop-nav"
        className={css`${navBarStyle}; ${mainNavBarStyle};`}>
            <Link href="/" passHref>
                <picture>
                    <Image 
                        data-testid="nav-logo"
                        width={120}
                        objectFit="contain"
                        src="/images/pokemon_logo.webp"
                            height={40} alt="Pokemon Logo"/>
                </picture>
            </Link>
            <div className={centerContentClass}>
                {menu.map(({Icon,path,title}) => (
                    <Link
                    key={path} href={path} passHref>
                        <div 
                        data-testid={`nav-item-${title}`}
                        className={css`${centerContentClass}; width: 40px; cursor: pointer`}>
                            <Icon color="white" size={28}/>
                        </div>
                    </Link>
                ))}
            </div>
        </nav>
    )
}