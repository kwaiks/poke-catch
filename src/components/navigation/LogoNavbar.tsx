import { css } from "@emotion/css"
import { SECONDARY_COLOR, MAIN_COLOR } from "config/constant"
import { navBarStyle } from "styles/emotion/list";
import Image from "next/image";

const mainNavBarStyle = css`
    color: white;
    background-color: ${MAIN_COLOR};
    justify-content: center;
`

export default function LogoNavbar({label}:{label?: string}){
    return (
        <nav 
        data-testid="logo-navbar"
        className={`${navBarStyle} ${mainNavBarStyle}`}>
            {
                label?
                <span>{label}</span> : 
                <picture>
                    <Image 
                    data-testid="nav-logo"
                    width={120}
                    objectFit="contain"
                    src="/images/pokemon_logo.webp"
                        height={40} alt="Pokemon Logo"/>
                </picture>
            }
        </nav>
    )
}