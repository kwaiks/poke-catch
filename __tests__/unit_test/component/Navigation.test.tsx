import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import BottomNavigation from 'components/navigation/BottomNavigation';
import DesktopNavBar from 'components/navigation/DesktopNavBar';
import LogoNavbar from 'components/navigation/LogoNavbar';
import menu from 'config/menu';
import { create } from 'react-test-renderer';
import BackNavbar from '../../../src/components/navigation/BackNavbar';

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

describe(("Navigation"), () => {
    useRouter.mockImplementation(() => ({}))

    describe(("Back Navbar"), () => {
        it("should render", () => {
            render(<BackNavbar/>)
    
            expect(screen.getByTestId("back-navbar")).toBeVisible();
            expect(screen.getByTestId("back-navbar-icon-wrapper")).toBeVisible();
            expect(screen.getByTestId("back-icon")).toBeVisible();
        });

        it("should match snapshot", () => {
            const wrapper = create(<BackNavbar/>).toJSON();

            expect(wrapper).toMatchSnapshot()
        })
    })

    describe(("Bottom Navigation"), () => {

        it("should render", () => {
            render(<BottomNavigation/>)
            const nav = screen.getByTestId("bottom-nav");
            expect(nav).toBeVisible();
        });

        menu.forEach((item) => {
            it("should render nav item", () => {
                render(<BottomNavigation/>)
                expect(screen.getByTestId(`nav-item-${item.title}`)).toBeVisible();
            })
        })

        it("should match snapshot", () => {
            const wrapper = create(<BottomNavigation/>).toJSON();

            expect(wrapper).toMatchSnapshot()
        })
    })

    describe(("Desktop Navbar"), () => {

        it("should render", () => {
            render(<DesktopNavBar/>);
            expect(screen.getByTestId("desktop-nav")).toBeVisible();
            expect(screen.getByTestId("nav-logo")).toBeVisible();
        });

        it("navigation match menus count", () => {
            render(<DesktopNavBar/>)
            const nav = screen.getByTestId("desktop-nav");
            expect(nav.childElementCount).toEqual(menu.length);
        })

        menu.forEach((item) => {
            it("should render nav item", () => {
                render(<DesktopNavBar/>)
                expect(screen.getByTestId(`nav-item-${item.title}`)).toBeVisible();
            })
        })

        it("should match snapshot", () => {
            const wrapper = create(<DesktopNavBar/>).toJSON();

            expect(wrapper).toMatchSnapshot()
        })
    })

    describe("Mobile Logo Navbar", () => {
        it("should render", ()=>{
            render(<LogoNavbar/>)
            expect(screen.getByTestId("logo-navbar")).toBeVisible();
            expect(screen.getByTestId("nav-logo")).toBeVisible()
        })

        it("should match snapshot", () => {
            const wrapper = create(<LogoNavbar/>).toJSON();

            expect(wrapper).toMatchSnapshot()
        })
    })

})
