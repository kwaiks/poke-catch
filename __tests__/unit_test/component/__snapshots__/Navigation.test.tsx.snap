// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Back Navbar should match snapshot 1`] = `
<nav
  className="css-18ttk36"
  data-testid="back-navbar"
>
  <div
    className="css-1x356my"
    data-testid="back-navbar-icon-wrapper"
    onClick={[Function]}
  >
    <svg
      data-testid="back-icon"
      focusable="false"
      height="20px"
      role="img"
      viewBox="0 0 320 512"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
        fill="#000"
      />
    </svg>
  </div>
</nav>
`;

exports[`Navigation Bottom Navigation should match snapshot 1`] = `
<div
  className="css-13m70it"
  data-testid="bottom-nav"
  id="footer"
>
  <div
    className="css-1o7and7"
    data-testid="nav-item-Home"
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <svg
      aria-hidden="true"
      height="20px"
      role="img"
      viewBox="0 0 576 512"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
        fill="#000"
      />
    </svg>
    <span>
      Home
    </span>
  </div>
  <div
    className="css-1o7and7"
    data-testid="nav-item-My Pokemon"
    href="/my-pokemon"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <svg
      height="20px"
      viewBox="0 0 1024 1024"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M 512.00,96.80 C 304.28,96.94 132.17,249.33 101.24,448.41 101.24,448.41 312.51,448.80 312.51,448.80 339.50,364.37 418.60,303.25 512.00,303.20 605.25,303.31 684.24,364.33 711.33,448.61 711.33,448.61 922.53,448.80 922.53,448.80 891.82,249.60 719.75,97.06 512.00,96.80 512.00,96.80 512.00,96.80 512.00,96.80 Z M 512.00,376.80 C 436.89,376.80 376.00,437.69 376.00,512.80 376.00,587.91 436.89,648.80 512.00,648.80 512.00,648.80 512.00,648.80 512.00,648.80 587.11,648.80 648.00,587.91 648.00,512.80 648.00,512.80 648.00,512.80 648.00,512.80 648.00,437.69 587.11,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 Z M 101.47,576.80 C 132.18,776.00 304.25,928.54 512.00,928.80 719.72,928.66 891.83,776.27 922.76,577.19 922.76,577.19 711.49,576.80 711.49,576.80 684.50,661.23 605.40,722.35 512.00,722.40 418.75,722.29 339.76,661.27 312.67,576.99 312.67,576.99 101.47,576.80 101.47,576.80 101.47,576.80 101.47,576.80 101.47,576.80 Z"
        fill="#000"
        id="pkmn-go-teamless"
        stroke="none"
      />
    </svg>
    <span>
      My Pokemon
    </span>
  </div>
</div>
`;

exports[`Navigation Desktop Navbar should match snapshot 1`] = `
<nav
  className="css-sdjexh"
  data-testid="desktop-nav"
>
  <picture
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "inline-block",
          "margin": 0,
          "maxWidth": "100%",
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "block",
            "maxWidth": "100%",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTIwIiBoZWlnaHQ9IjQwIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
          style={
            Object {
              "border": "none",
              "display": "block",
              "margin": 0,
              "maxWidth": "100%",
              "padding": 0,
            }
          }
        />
      </div>
      <noscript>
        <img
          alt="Pokemon Logo"
          data-testid="nav-logo"
          decoding="async"
          src="/_next/image?url=%2Fimages%2Fpokemon_logo.webp&w=256&q=75"
          srcSet="/_next/image?url=%2Fimages%2Fpokemon_logo.webp&w=128&q=75 1x, /_next/image?url=%2Fimages%2Fpokemon_logo.webp&w=256&q=75 2x"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": "contain",
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </noscript>
      <img
        alt="Pokemon Logo"
        data-testid="nav-logo"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": "contain",
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
    </div>
  </picture>
  <div
    className="css-1ieito0"
  >
    <div
      className="css-1b4f159"
      data-testid="nav-item-Home"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        aria-hidden="true"
        height={28}
        role="img"
        viewBox="0 0 576 512"
        width={28}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          fill="white"
        />
      </svg>
    </div>
    <div
      className="css-1b4f159"
      data-testid="nav-item-My Pokemon"
      href="/my-pokemon"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        height={28}
        viewBox="0 0 1024 1024"
        width={28}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M 512.00,96.80 C 304.28,96.94 132.17,249.33 101.24,448.41 101.24,448.41 312.51,448.80 312.51,448.80 339.50,364.37 418.60,303.25 512.00,303.20 605.25,303.31 684.24,364.33 711.33,448.61 711.33,448.61 922.53,448.80 922.53,448.80 891.82,249.60 719.75,97.06 512.00,96.80 512.00,96.80 512.00,96.80 512.00,96.80 Z M 512.00,376.80 C 436.89,376.80 376.00,437.69 376.00,512.80 376.00,587.91 436.89,648.80 512.00,648.80 512.00,648.80 512.00,648.80 512.00,648.80 587.11,648.80 648.00,587.91 648.00,512.80 648.00,512.80 648.00,512.80 648.00,512.80 648.00,437.69 587.11,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 512.00,376.80 Z M 101.47,576.80 C 132.18,776.00 304.25,928.54 512.00,928.80 719.72,928.66 891.83,776.27 922.76,577.19 922.76,577.19 711.49,576.80 711.49,576.80 684.50,661.23 605.40,722.35 512.00,722.40 418.75,722.29 339.76,661.27 312.67,576.99 312.67,576.99 101.47,576.80 101.47,576.80 101.47,576.80 101.47,576.80 101.47,576.80 Z"
          fill="white"
          id="pkmn-go-teamless"
          stroke="none"
        />
      </svg>
    </div>
  </div>
</nav>
`;

exports[`Navigation Mobile Logo Navbar should match snapshot 1`] = `
<nav
  className="css-1l0d73p css-3wreig"
  data-testid="logo-navbar"
>
  <picture>
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "inline-block",
          "margin": 0,
          "maxWidth": "100%",
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "block",
            "maxWidth": "100%",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTIwIiBoZWlnaHQ9IjQwIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
          style={
            Object {
              "border": "none",
              "display": "block",
              "margin": 0,
              "maxWidth": "100%",
              "padding": 0,
            }
          }
        />
      </div>
      <noscript>
        <img
          alt="Pokemon Logo"
          data-testid="nav-logo"
          decoding="async"
          src="/_next/image?url=%2Fimages%2Fpokemon_logo.webp&w=256&q=75"
          srcSet="/_next/image?url=%2Fimages%2Fpokemon_logo.webp&w=128&q=75 1x, /_next/image?url=%2Fimages%2Fpokemon_logo.webp&w=256&q=75 2x"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": "contain",
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </noscript>
      <img
        alt="Pokemon Logo"
        data-testid="nav-logo"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": "contain",
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
    </div>
  </picture>
</nav>
`;
